[JIT] integer value refinement

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56966 [JIT] Disable Complete Shape Inlining For Testing Purposes
* #56828 Add unary/binary ops necessary for mobilenet
* #56642 Only optimize while change has been made
* **#56438 [JIT] integer value refinement**
* #56437 Factor out isDominatedBy
* #55978 Add Value * == Value * peephole
* #55977 [JIT] [Easy] Factor out non tensor peephole
* #55926 Add list len refinement
* #55925 Add handling of symbolic shapes
* #54809 Initial Symbolic Shape Analysis

Differential Revision: [D27924239](https://our.internmc.facebook.com/intern/diff/D27924239)

Commit:
temp

[ghstack-poisoned]

Commit:
Update on "temp"




[ghstack-poisoned]

Commit:
Update on "temp"




[ghstack-poisoned]

Commit:
Update on "[JIT] integer value refinement"




[ghstack-poisoned]

Commit:
Update on "[JIT] integer value refinement"




[ghstack-poisoned]

Commit:
Update on "[JIT] integer value refinement"




[ghstack-poisoned]

Commit:
Update on "[JIT] integer value refinement"




[ghstack-poisoned]

Commit:
Update on "[JIT] integer value refinement"


Differential Revision: [D27924239](https://our.internmc.facebook.com/intern/diff/D27924239)

[ghstack-poisoned]

Commit:
Update on "[JIT] integer value refinement"


Differential Revision: [D27924239](https://our.internmc.facebook.com/intern/diff/D27924239)

[ghstack-poisoned]

penguinwu commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,199 @@
+    return false;
+  }
+
+  ListRefinement RefineListLens(Block* b, ListRefinement block_refinements) {
```
Are we refining list lens or integers? If it is the latter, `RefineListLens` is misleading.

penguinwu replied:
What is `ListRefinement block_refinements` here? How is it different from `IntegerRefinement`?

penguinwu replied:
Please add comments on what `RefineListLens` do to help to understand (especially because it is called recursively).

penguinwu commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,199 @@
+using IntegerRefinement = std::unordered_map<Value*, int64_t>;
+
+// see [value refinement algorithm]
+
```
Even though we have a pointer to the value refinement algorithm, it would be easier to understand to add a simple example of IntegerValueRefiner as the comment here.

penguinwu commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,199 @@
+  // A stack of active refinements, one for each block
+  std::vector<ListRefinement*> active_refinements_;
+  // A map from Boolean Value * -> associated refinements
+  std::unordered_map<Value*, BooleanRefinementMapping> info_;
```
`info_` is too generic a name, how about `bool_values_to_refinement_map` for readability?

penguinwu commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,199 @@
+        // both block outputs with the same value, which opens up
+        // further optimization opportunities
+        // The pass will already handle if both are refined to the same
+        // constant. Here, we add a case case where one block output
```
"case case" => "case

penguinwu commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,199 @@
+              continue;
+            }
+            // Value must be in scope for both blocks
+            if (!if_n.node()->isDominatedBy(block_output->node())) {
```
`if_n` is the if-node, `block_output` is `if_block->outputs().at(i)` where `if_block` is `if_n.node()->blocks().at(block_index)`. Isn't `if_n` always dominated by `block_output`? Adding an example would help here.

eellison replied:
the one block output doesn't necessarily dominate the other block output

penguinwu replied:
I see. I thought if-node output is always a phi-node that merges values from both branches thus would always dominate the if-node. People may get confused if they do not know the IR structure intimately, adding an example would help a lot here.

penguinwu commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,199 @@
+            if_n.elseBlock(),
+            has_cond_ref ? info_[if_n.cond()].false_refine() : empty);
+
+        // we are looking for cases where we can replace
```
I feel carving this special handling out into a separate function would help the readability of the algorithm.

penguinwu commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,199 @@
+
+    // this is useful for things like if block outputs,
+    // where the output value node may not be defined in this block
+    // but we have refined its use as a block output
```
Add an example would help

penguinwu reviewed (changes requested):
I need the requested comments/examples to understand the algorithm better, thus request change.

A bigger concern I have is that the algorithm is non-trivial to understand and bears a lot of similarity with ListLenRefinement. For the sake of maintainability, can we combine all value refinement into one code to reuse as much code as possible? I am fine w/ doing it as a separate refactoring PR.

Combining the common code paths would also make the structure of the algorithm cleaner. Bugs exposed along one type of refinement are likely bugs on another type of refinement so maximize reuse would make it a lot easier to expose and fix common bugs.

eellison commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,199 @@
+        throwing_blocks_.insert(b);
+      }
+
+      for (size_t input = 0; input < n->inputs().size(); ++input) {
```
this does not apply to the other optimization, there we are optimizing `aten::len`, here we are optimizing all uses

eellison commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,199 @@
+  ListRefinement RefineListLens(Block* b, ListRefinement block_refinements) {
+    active_refinements_.push_back(&block_refinements);
+    for (Node* n : b->nodes()) {
+      if (n->matches("aten::eq(int a, int b) -> bool") ||
```
this also does not apply, as we are pattern matching a separate node

eellison commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,199 @@
+        // can always safely be replaced with `y`
+        // this is an important case for symbolic shape analysis
+
+        for (size_t block_index : {0, 1}) {
```
this is optimization doesn't apply to the other optimization, since we are refining uses here and not just `aten::len`, so you will not have a similar case in the other optimization

eellison commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,199 @@
+    // this is useful for things like if block outputs,
+    // where the output value node may not be defined in this block
+    // but we have refined its use as a block output
+    for (size_t i = 0; i < b->outputs().size(); ++i) {
```
Also does not apply - optimizing uses not just `aten::len`

eellison reviewed (commented):
@penguinwu in my opinion all of the common code is already abstracted out into a helper functions. I personally don't find abstracted classes to be any easier to understand than factoring out through composition (functions). Additionally, we are very likely going to add a Tensor refinement pass. I don't want to add additional layers of abstraction beyond factoring out common code when we are still unclear what the right abstraction may be yet, because the Tensor refinement use case will be a much more different use case than integer value refinement is to len refinement.

I will add more comments and improve naming like I did for other PR - sorry probably should have updated to match with other PR before asking for review

eellison committed:
Update on "[JIT] integer value refinement"


Differential Revision: [D27924239](https://our.internmc.facebook.com/intern/diff/D27924239)

[ghstack-poisoned]

eellison committed:
Update on "[JIT] integer value refinement"


Differential Revision: [D27924239](https://our.internmc.facebook.com/intern/diff/D27924239)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27924239).

ZolotukhinM commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,215 @@
+    return changed_;
+  }
+
+  bool checkForPossibleRefinements(Block* b) {
```
Nit: name can be made more specific, like "blockHasIntComparisons". Up to you of course, but "possible refinements" might sound a bit too vague.

eellison replied:
`blockHasIntComparisons` does describe the current pass, but it also doesn't leave room for improvement, e.g. I could add checks that either value in the int comparison has subsequent uses

ZolotukhinM replied:
When an improvement is made, we can rename it to reflect the new functionality :)

ZolotukhinM commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,215 @@
+      Node* if_node,
+      IntegerRefinement& true_block_refinements,
+      IntegerRefinement& false_block_refinements) {
+    // we are looking for cases where we can replace
+    // both block outputs with the same value, which opens up
+    // further optimization opportunities
+    // The pass will already handle if both are refined to the same
+    // constant. Here, we add a case case where one block output
```
Nit: formatting looks odd. Also, some phrases seem unfinished?

eellison replied:
this is just default clang-format :/ if you join the lines together the sentences makes sense

ZolotukhinM commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,215 @@
+    // constant. Here, we add a case case where one block output
+    // is refined to a constant in the other block, and where the existing
+    // block output in the other block is the same constant
+    // x = 1
+    // if y == 1:
+    //    return x
+    // else:
+    //    return y
```
It would be easier to understand code if you included TS IR dumps as well. It's a contiguous mental overhead to translate this into an IR in your head as you read the code (i.e. what's `block_output` in this example?)

ZolotukhinM commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,215 @@
+        }
+        const auto& other_block_refinements =
+            block_index == 0 ? false_block_refinements : true_block_refinements;
+        c10::optional<int64_t> maybe_refine = tryFindRefinement(block_output);
```
This code in particular is very hard to reason about without an example at hand. I would recommend adding example + adding comments explaining what each value in this code is holding in that example. E.g.:
```
block_output = %y.1
maybe_refine = {%x.1 --> 1} // the actual value should probably be different here
other_block_refinements = {%y.1 --> 1}
```

eellison replied:
how is this different than the example given above ?

ZolotukhinM replied:
The example above doesn't show what each of these variables hold (other_block_refinements, block_output, ...) - what I want here is almost a walkthrough the code reflected in the comments, e.g. "after this line 'block_output' holds this", "after that line maybe_refine gets this value".

Also, I'd like to see the example formulated in terms of TorchScript so that it's easier to follow (as I mentioned in the previous comment).

ZolotukhinM commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    // this is useful for things like if block outputs,
+    // where the output value node may not be defined in this block
```
Nit: what does "output value node may not be defined" mean?

eellison replied:
Will clarify..

ZolotukhinM replied:
To clarify a bit: I think the word 'node' simply needs to be deleted there :)

ZolotukhinM commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,215 @@
+    // where the output value node may not be defined in this block
+    // but we have refined its use as a block output
+    for (size_t i = 0; i < b->outputs().size(); ++i) {
+      Value* input_v = b->outputs().at(i);
```
Why is it `input_v` and not `output_v`?

ZolotukhinM reviewed (commented):
I have mostly readability concerns. I'd like to have one more quick look after they are addressed to make sure I can understand the code.

eellison committed:
Update on "[JIT] integer value refinement"


Differential Revision: [D27924239](https://our.internmc.facebook.com/intern/diff/D27924239)

[ghstack-poisoned]

eellison commented:
@ZolotukhinM thanks for your comments! agreed they made the code more readable.

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27924239).

eellison commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -0,0 +1,227 @@
+    //     block1():
+    //       -> (%y.1)
+    //   return (%15)
+    // Here, %15 can always be safely replaced with %y.1
```
No, because if `%3` is false `%15` will be equal to `%y.1`

eellison committed:
Update on "[JIT] integer value refinement"


Differential Revision: [D27924239](https://our.internmc.facebook.com/intern/diff/D27924239)

[ghstack-poisoned]

eellison committed:
Update on "[JIT] integer value refinement"


Differential Revision: [D27924239](https://our.internmc.facebook.com/intern/diff/D27924239)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27924239).

eellison committed:
Update on "[JIT] integer value refinement"


Differential Revision: [D27924239](https://our.internmc.facebook.com/intern/diff/D27924239)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27924239).

eellison committed:
Update on "[JIT] integer value refinement"


Differential Revision: [D27924239](https://our.internmc.facebook.com/intern/diff/D27924239)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27924239).
